tests/ts/logger_helpers.ts

```diff
@@ -0,0 +1,511 @@
+            const logger = new TestLogger();
+            const root = new LevelLogger(logger, 'debug');
+
+            // Ensure a record of each expected type is being logged
+            logEachType(root);
+            expectLogged({ level: 'debug', logger: logger });
+        });
+
+        it("'info' level discards 'debug' and 'trace'", () => {
+            const logger = new TestLogger();
+            const root = new LevelLogger(logger, 'info');
+
+            // Ensure a record of each expected type is being logged
+            logEachType(root);
+            expectLogged({ level: 'info', logger: logger });
+        });
+
+        it("'warn' level discards 'debug', 'trace' and 'info'", () => {
+            const logger = new TestLogger();
+            const root = new LevelLogger(logger, 'warn');
+
+            // Ensure a record of each expected type is being logged
+            logEachType(root);
+            expectLogged({ level: 'warn', logger: logger });
+        });
+
+        it("'error' level discards 'debug', 'trace', 'info' and 'warn'", () => {
+            const logger = new TestLogger();
+            const root = new LevelLogger(logger, 'error');
+
+            // Ensure a record of each expected type is being logged
+            logEachType(root);
+            expectLogged({ level: 'error', logger: logger });
+        });
+    });
+
+    describe('TagLogger', () => {
+        it('applies a tag', () => {
+            const logger = new TestLogger();
+            const root = new TagLogger(logger, 'tag');
+
+            // Ensure a record of each type is being logged with the expected tag
+            logEachType(root);
+            expectLogged({ level: 'debug', tag: 'tag', logger: logger });
+        });
+
+        it('applies multiple tags', () => {
+            const logger = new TestLogger();
+            const root = new TagLogger(logger, 'tag1', 'tag2');
+
+            // Ensure a record of each type is being logged with the expected tag
+            logEachType(root);
+            expectLogged({ level: 'debug', tag: ['tag1', 'tag2'], logger: logger });
+        });
+    });
+
+    describe('UnveilLogger', () => {
+        it('leaves normal log records untouched', () => {
+            const logger = new TestLogger();
+            const root = new UnveilLogger(logger);
+
+            // Ensure a record of each expected type is being logged
+            logEachType(root);
+            expectLogged({ level: 'debug', logger: logger });
+        });
+
+        it('unveils confidential log record', () => {
+            const logger = new TestLogger();
+            const root = new UnveilLogger(logger);
+
+            // Ensure a confidential record is being unveiled
+            root.debug(new TestConfidential());
+            expect(logger.records).toContain(['debug', 'uncensored']);
+        });
+    });
+
+    describe('ConsoleLogger', () => {
+        let backup: Logger = {} as Logger;
```

Comment:
Why do you cast this to `Logger`? Isn't this a plain `object` containing loggers as values?

Reply:
This holds the `Console` log level methods until they are being restored. I could leave it as an object but since it actually does implement the interface, it doesn't hurt.

Reply:
Ahh, I didn't relalize that it *does* implement the interface. In that case it makes sense of course :slightly_smiling_face: